### ISSUE TITLE
Diff not behaving as expected - key not being assigned to components

### ISSUE THREAD

**Issue by lerran80** [developit/preact#655]
In a structure like this:
http://codepen.io/LC80/pen/mmVWpE?editors=0010

Where a list of components is rendered from a mapped array, connected to a redux store.

If something that does not affect the Parent component props is changed in the store, the Parent component is re-rendered anyway. 
   - If the mapped children are inside a <div>, diff works as expected and they are not re-rendered to the DOM
   - If the key is passed directly to the children components, diff apparently does no work, all children are re-rendered to the DOM. Devtools shows that the key is not assigned to the child component in this case.

Not sure if my explanation was sufficiently clear, it may be easier checking my comments in the codepen.

BTW: preact is a life saver in places like Brazil, where user bandwidth is a severe constraint. Thank you so much for this awesome work!

**Comment by thysultan** [comment]
This might be the same problem as (https://github.com/developit/preact/issues/642)[https://github.com/developit/preact/issues/642] since preact normalizes nested arrays in children you have a case where the two `.map` constructs are producing children with similar keys and react might be adding a depth to the key to identify it's position when nested.

**Comment by developit** [comment]
It's likely our devtools integration is hiding `key` on components, so that might be unrelated here (but we should definitely get that fixed).

The issue here is actually that you have duplicate keys, because you're looping over `labels` twice within the same parent element (the wrapping `<div>`):

```js
<div>
  {labels.map(i => <div key={i}>{h(Child, {i})}</div>) }
  {labels.map(i => h(Child,{key:i, i}))}
</div>

// the above produces:
<div>
  <div key="a"><Child i="a" /></div>
  <div key="b"><Child i="b" /></div>
  <div key="c"><Child i="c" /></div>

  <Child key="a" i="a" />
  <Child key="b" i="b" />
  <Child key="c" i="c" />
</div>
```

Preact's handling of duplicate keys is a little odd right now, we're looking at fixing that in #642 as @thysultan noted.  We've also added duplicate key warnings to `preact/debug`, which will be released shortly. (maybe today!)

Here's your pen, but updated to prefix the keys in the first loop with "a" and the second loop with "b" so that they don't collide:

http://codepen.io/developit/pen/WjrEbp?editors=0010

**Comment by lerran80** [comment]
Thanks for the quick response!

Actually, I found this problem in my codebase, which is too large and complex to use as an example. So I created this pen to make it easier to show you guys. I just checked in my original code, there are no duplicate keys.

I changed the codepen (same one: http://codepen.io/LC80/pen/mmVWpE?editors=0011 ) to better illustrate what I am seeing. As you can see in the console, all children are being re-rendered in both cases, whenever state changes. I wrongly assumed it was just the second group not being re-rendered because of some side effects in my app, sorry about the confusion.

I tried to access the key prop inside the children, I would expect to see it for the second group, but they are undefined. Not sure if they were supposed to be passed down.

**Comment by developit** [comment]
Ah, so that's a whole different issue haha.

In preact, `key` is not passed to components as a prop. As far as I'm aware this is true in React as well.  You can see the special-case for key and ref props here:
https://github.com/developit/preact/blob/master/src/vdom/component.js#L20-L21

**Comment by lerran80** [comment]
Oh, ok. No problem there! I was just trying to access it for debugging. The main issue is that the mapped children are being re-rendered, even if they have no duplicate keys.

After you type something in the input box, you will see " State changed!" in console, followed by all the "Rendered ..." logs of all child components. As they have a stable key, I was not expecting those extra logs, as they would not be re-rendered.

**Comment by developit** [comment]
Adding a key does not prevent rendering, only `shouldComponentUpdate` can do that.  They work nicely in combination though:

```js
const List = ({ items }) => (
  items.map( item => (
    <ListItem key={item} item={item} />
  ))
)

class ListItem extends Component {
  shouldComponentUpdate(nextProps) {
    // only re-render if passed a new item
    return nextProps.item !== this.props.item;
  }
  render({ item }) {
    return <span>{item}</span>
  }
}

// Demo:
let root;

// renders all 3 items:
render(<List items={['a', 'b', 'c']} />, document.body, root);

// just removes 'b' and 'c', does not diff 'a'
render(<List items={['a']} />, document.body, root);

// first ('a') is skipped. 'b', 'c' and 'd' are added.
render(<List items={['a', 'b', 'c', 'd']} />, document.body, root);

// thanks to keys, we can shuffle the list and it'll move the children without re-rendering:
render(<List items={['b', 'd', 'c', 'a']} />, document.body, root);
```

**Comment by lerran80** [comment]
Ah, right. So, I misinterpreted how key is used. Thanks for the explanation.

If I understood correctly, the key will improve re-render performance but not by preventing child re-renders from happening. My impression is that your example would still work without the key prop, but would not have optimal performance, correct?

**Comment by developit** [comment]
`key` just explicitly tells the diff which element is which in 2 renders.  It's easiest to see with your list as an example - imagine the following:

```js
// first render returns 3 items:
<li>a</li><li>b</li><li>c</li>

// second render removes the second item:
<li>a</li><li>c</li>
```

Since render() just returns a new virtual DOM tree each time, there's no way for a library to know that the second item (`b`) was removed.  By default, it will actually assume the last item was removed, and "reflow" all of the elements to drop the last one off.  By adding a unique `key` to each element that is consistent between renders, you're explicitly telling the diff which element is which:

```js
<li key="a">a</li><li key="b">b</li><li key="c">c</li>

<li key="a">a</li><li key="c">c</li>
```

In the above, we can see that the diff will compare the two lists and determine `key="b"` is not present in the second list (the second render).  Since it's missing, it'll remove that element and the first and last elements are left untouched.

I probably did a horrible job explaining that, sorry!  I did have a StackOverflow post about keys a while back that people seemed to like:
http://stackoverflow.com/a/42775667/194340
